### ISSUE TITLE
Training instability

### ISSUE THREAD

**Issue by Taoboan1999** [ultralytics/yolov5#13608]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I'm training yolov12's model on my own dataset, and on map50 of the validation set, I often get 0.67→0.68→0.69→0.70→0.71→0.68→0.67→0.67→0.68→0.67→0.67
Such approximate changes in results. I ensured maximum batchsize=8 and maximum image size, tried reducing the learning rate lr0=0.001→0.0001→0.00001, and increased the parameters for data augmentation (mosaic).

I can't change all the data in my training set and validation set, how can I change this bad status quo please

Translated with DeepL.com (free version)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Taoboan1999, thank you for reaching out and for your detailed description! 🚀 This is an automated response to help you get started, and an Ultralytics engineer will assist you soon.

Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) for guidance, including quickstart guides for [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) and tips for [Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

If this is a 🐛 Bug Report, please provide a **minimum reproducible example (MRE)** to help us debug more effectively.

If your question is about custom training, please include as much information as possible—such as dataset samples, training logs, and the exact commands you are using.

## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```
## Environments
YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience! An Ultralytics engineer will review your issue and provide more specific guidance soon. 😊

**Comment by pderrenger** [comment]
Hi @Taoboan1999! The training instability you're experiencing with oscillating mAP50 values (0.67-0.71) is common and can be addressed with several adjustments. Try implementing learning rate scheduling with `--cos-lr` for cosine annealing, reduce `--momentum` from default 0.937 to 0.9, and consider using `--patience` for early stopping to prevent overfitting. You might also benefit from using [Exponential Moving Average (EMA)](https://docs.ultralytics.com/guides/model-training-tips/#exponential-moving-average-ema) which smooths model weights during training and often leads to more stable convergence patterns.

**Comment by Taoboan1999** [comment]
> Hi [@Taoboan1999](https://github.com/Taoboan1999)! The training instability you're experiencing with oscillating mAP50 values (0.67-0.71) is common and can be addressed with several adjustments. Try implementing learning rate scheduling with `--cos-lr` for cosine annealing, reduce `--momentum` from default 0.937 to 0.9, and consider using `--patience` for early stopping to prevent overfitting. You might also benefit from using [Exponential Moving Average (EMA)](https://docs.ultralytics.com/guides/model-training-tips/#exponential-moving-average-ema) which smooths model weights during training and often leads to more stable convergence patterns.

Thank you very much for your patience, I will try the various methods you mentioned one by one and if any of them are more effective, I will post them

**Comment by Taoboan1999** [comment]
> Hi [@Taoboan1999](https://github.com/Taoboan1999)! The training instability you're experiencing with oscillating mAP50 values (0.67-0.71) is common and can be addressed with several adjustments. Try implementing learning rate scheduling with `--cos-lr` for cosine annealing, reduce `--momentum` from default 0.937 to 0.9, and consider using `--patience` for early stopping to prevent overfitting. You might also benefit from using [Exponential Moving Average (EMA)](https://docs.ultralytics.com/guides/model-training-tips/#exponential-moving-average-ema) which smooths model weights during training and often leads to more stable convergence patterns.

How do you turn on EMA in yolov12 training.

**Comment by pderrenger** [comment]
Hi @Taoboan1999! EMA is enabled by default in YOLOv5 training - you can control it with the `--ema` flag (enabled by default) and adjust the decay rate using `--ema-decay` (default 0.9999). If you want to disable EMA for testing, use `--ema False`, or to modify the decay rate for potentially more stable training, try `--ema-decay 0.999` for faster adaptation to recent weights.

**Comment by Taoboan1999** [comment]
> Hi [@Taoboan1999](https://github.com/Taoboan1999)! EMA is enabled by default in YOLOv5 training - you can control it with the `--ema` flag (enabled by default) and adjust the decay rate using `--ema-decay` (default 0.9999). If you want to disable EMA for testing, use `--ema False`, or to modify the decay rate for potentially more stable training, try `--ema-decay 0.999` for faster adaptation to recent weights.

Thank you again for your answer. I will try modifying this parameter when training YOLOv5, but I am still encountering overfitting issues in YOLOv12. I will ask a question in the YOLOv12 issues section. Thank you.

**Comment by pderrenger** [comment]
You're welcome @Taoboan1999! Just to clarify, this is the YOLOv5 repository, so for YOLOv12 (YOLO12) questions you'll want to head over to the [main Ultralytics repository](https://github.com/ultralytics/ultralytics) where the newer YOLO versions are maintained. Good luck with your training!
